### PR TITLE
etcd, pd (ticdc): refine pdClient and etcdClient initialization

### LINKED ISSUE #9565: two of ticdc restart repeatedly after injection network partition between pdleader and pd followers
### What did you do?

1、run tpcc with 1000 warehouse and 10 thread
2、inject network partition between pdleader and pd followers last for 10mins
3、after 10mins，recover fault

### What did you expect to see?

all ticdc are normal

### What did you see instead?

one of ticdc restart repeatedly even after fault recovery

image

[2023/08/11 19:46:05.183 +08:00] [INFO] [capture.go:308] ["the capture routine has exited"]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:214] ["exit tso dispatcher loop"]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:375] ["[tso] stop fetching the pending tso requests due to context canceled"] [dc-location=global]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:162] ["exit tso requests cancel loop"]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:162] ["exit tso requests cancel loop"]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:311] ["[tso] exit tso dispatcher"] [dc-location=global]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:214] ["exit tso dispatcher loop"]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:375] ["[tso] stop fetching the pending tso requests due to context canceled"] [dc-location=global]
[2023/08/11 19:46:05.183 +08:00] [INFO] [tso_dispatcher.go:311] ["[tso] exit tso dispatcher"] [dc-location=global]
[2023/08/11 19:46:05.183 +08:00] [ERROR] [server.go:286] ["http server error"] [error="[CDC:ErrServeHTTP]serve http error: mux: server closed"] [errorVerbose="[CDC:ErrServeHTTP]serve http error: mux: server closed\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/errors.WrapError\n\tgithub.com/pingcap/tiflow/pkg/errors/helper.go:34\ngithub.com/pingcap/tiflow/cdc/server.(*server).startStatusHTTP.func1\n\tgithub.com/pingcap/tiflow/cdc/server/server.go:286\nruntime.goexit\n\truntime/asm_amd64.s:1650"]
[2023/08/11 19:46:05.183 +08:00] [WARN] [server.go:140] ["cdc server exits with error"] [error="[CDC:ErrReachMaxTry]reach maximum try: 10, error: [CDC:ErrCheckClusterVersionFromPD]failed to request PD 503 Service Unavailable no leader\n, please try again later: [CDC:ErrCheckClusterVersionFromPD]failed to request PD 503 Service Unavailable no leader\n, please try again later"] [errorVerbose="[CDC:ErrReachMaxTry]reach maximum try: 10, error: [CDC:ErrCheckClusterVersionFromPD]failed to request PD 503 Service Unavailable no leader\n, please try again later: [CDC:ErrCheckClusterVersionFromPD]failed to request PD 503 Service Unavailable no leader\n, please try again later\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164\ngithub.com/pingcap/tiflow/pkg/retry.run\n\tgithub.com/pingcap/tiflow/pkg/retry/retry_with_opt.go:69\ngithub.com/pingcap/tiflow/pkg/retry.Do\n\tgithub.com/pingcap/tiflow/pkg/retry/retry_with_opt.go:34\ngithub.com/pingcap/tiflow/pkg/version.CheckClusterVersion\n\tgithub.com/pingcap/tiflow/pkg/version/check.go:93\ngithub.com/pingcap/tiflow/pkg/upstream.initUpstream\n\tgithub.com/pingcap/tiflow/pkg/upstream/upstream.go:157\ngithub.com/pingcap/tiflow/pkg/upstream.(*Manager).AddDefaultUpstream\n\tgithub.com/pingcap/tiflow/pkg/upstream/manager.go:92\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).reset\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:230\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).run\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:313\ngithub.com/pingcap/tiflow/cdc/capture.(*captureImpl).Run\n\tgithub.com/pingcap/tiflow/cdc/capture/capture.go:288\ngithub.com/pingcap/tiflow/cdc/server.(*server).run.func1\n\tgithub.com/pingcap/tiflow/cdc/server/server.go:345\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.2.0/errgroup/errgroup.go:75\nruntime.goexit\n\truntime/asm_amd64.s:1650"]
[2023/08/11 19:46:05.183 +08:00] [INFO] [capture.go:688] ["message router closed"] [captureID=6fba15d5-480c-43a7-9dd0-121b539fd12f]
[2023/08/11 19:46:05.183 +08:00] [INFO] [server.go:412] ["sort engine manager closed"] [duration=7.455µs]
[2023/08/11 19:47:35.650 +08:00] [INFO] [helper.go:54] ["init log"] [file=/var/lib/ticdc/log/ticdc.log] [level=info]

### Versions of the cluster

[2023/08/11 19:47:35.650 +08:00] [INFO] [version.go:47] ["Welcome to Change Data Capture (CDC)"] [release-version=v7.4.0-alpha] [git-hash=0eef200142f4922736cf9e5ea1067a4d6c617329] [git-branch=heads/refs/tags/v7.4.0-alpha] [utc-build-time="2023-08-11 11:36:27"] [go-version="go version go1.21.0 linux/amd64"] [failpoint-build=false]

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

_No response_

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #9565  #9532 
ref #9054 

### What is changed and how it works?
This PR can decrease the probability of cdc capture restarting during pd network isolation and increase the success rate of restarts, resulting in reduced delay for changefeed in this scenario. I tested it on tisapce and found that with this PR, network isolation no longer causes cdc capture and server to exit. (In theory, it is still possible for cdc to exit, but I did not observe it in my testing. Moreover, if cdc exits, the success rate of restarts will be greatly improved.)

The specific approach is:

1. Reduce the number of pdClients that need to be created when cdc server starts from 4 to 2. This can decrease the probability of cdc server failing to start due to pdClient creation failure.
2. After successfully creating the first pdClient, use it to obtain all online pd addresses and update the pd addresses passed to server startup parameters. After this, new addresses will be used for creating other pdClients and etcdClients.
3. When creating etcdClient, a goroutine will be started to maintain its internal address update logic, which is basically copied from pd. Reference: https://github.com/tikv/pd/blob/master/pkg/utils/etcdutil/etcdutil.go#L249

#### Tests 

 - Manual test (add detailed scripts or steps below)
1. Deploy a TiDB cluster with 3 PD nodes and 2 TiCDC nodes, create a synchronization task, and continuously write to the upstream TiDB.
2. Use the following iptables command to inject network isolation into any PD node.

```bash
sudo iptables -A OUTPUT -p tcp --dport 2379 -j DROP

```

3. During the test, observe the server status and synchronization delay of CDC, both of which are normal.

### Release note 

```release-note
Fix a bug that may cause cdc restart many time when there is a PD network isolation in the upstream PD cluster.
```

### CODE REVIEW

File: pkg/util/comparison.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import "sort"
+
+// AreStringSlicesEquivalent checks if two string slices are equivalent.
+// If the slices are of the same length and contain the same elements (but possibly in different order), the function returns true.
+func AreStringSlicesEquivalent(a, b []string) bool {
```
Comment (sdojjy) [inline pkg/util/comparison.go:20]: this function changes the original string slice order, please add comments.

File: pkg/util/comparison_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+func TestAreStringSlicesEquivalent(t *testing.T) {
+	tests := []struct {
+		name string
+		a    []string
+		b    []string
+		want bool
+	}{
+		{
```
Comment (sdojjy) [inline pkg/util/comparison_test.go:26]: add two nil slice case

**Review by sdojjy — APPROVED** [review]

File: cdc/capture/capture.go [hunk 3]
```diff
@@ -81,6 +82,7 @@ type captureImpl struct {
 	liveness         model.Liveness
 	config           *config.ServerConfig
 
+	pdClient        pd.Client
 	pdEndpoints     []string
```
Comment (3AceShowHand) [inline cdc/capture/capture.go:86]: `pdClient` and `pdEndpoint` are maintained at the same time, this looks weird

File: pkg/etcd/client.go [hunk 4]
```diff
@@ -313,3 +322,229 @@ func isRetryableError(rpcName string) retry.IsRetryable {
...
+		TLS:                  tlsConfig,
+		LogConfig:            &logConfig,
+		DialTimeout:          defaultEtcdClientTimeout,
+		DialKeepAliveTime:    defaultDialKeepAliveTime,
+		DialKeepAliveTimeout: defaultDialKeepAliveTimeout,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return client, err
```
Comment (3AceShowHand) [inline pkg/etcd/client.go:364]: `return client, nil`

File: pkg/etcd/client.go [hunk 5]
```diff
@@ -313,3 +322,229 @@ func isRetryableError(rpcName string) retry.IsRetryable {
...
+	tickerInterval := defaultDialKeepAliveTime
+
+	checker := &healthyChecker{
+		tlsConfig: tlsConfig,
+	}
+	eps := syncUrls(client)
+	checker.update(eps)
+
+	// Create a goroutine to check the health of etcd endpoints periodically.
+	go func(client *clientv3.Client) {
```
Comment (3AceShowHand) [inline pkg/etcd/client.go:384]: this spawns 2 goroutines which are not managed by any synchronization mechanism, such as the `WaitGroup`

**Review by 3AceShowHand — APPROVED** [review]

File: pkg/etcd/client.go (hunk 5 above)
Comment (asddongmen) [reply pkg/etcd/client.go:384]: These two goroutines have the same lifecycle as the CDC server process, so there is no need to add any additional controls for them.

**Pull request merged**